File: apps/api/src/routes/billing.ts
import express, { Request, Response } from 'express';
import crypto from 'crypto';
import Razorpay from 'razorpay';
import { db } from '@zenvort/db';

const router = express.Router();

const CREDIT_PACKS = {
  starter: { credits: 500, amount: 199, name: 'Starter Pack' },
  pro: { credits: 2000, amount: 599, name: 'Pro Pack' },
  enterprise: { credits: 10000, amount: 1999, name: 'Enterprise Pack' },
} as const;

type PackType = keyof typeof CREDIT_PACKS;

// Auth middleware helper
async function requireAuth(req: Request): Promise<any> {
  const authHeader = req.headers.authorization;
  if (!authHeader || !authHeader.startsWith('Bearer ')) {
    throw { status: 401, message: 'Missing API key' };
  }
  const apiKey = authHeader.split(' ')[1];
  const user = await db.user.findUnique({ where: { apiKey } });
  if (!user) throw { status: 401, message: 'Invalid API key' };
  return user;
}

// GET /billing/plans
router.get('/plans', (_req: Request, res: Response) => {
  const plans = Object.entries(CREDIT_PACKS).map(([key, pack]) => ({
    pack: key,
    credits: pack.credits,
    amount: pack.amount,
    currency: 'INR',
    name: pack.name,
  }));
  return res.json(plans);
});

// GET /billing/usage
router.get('/usage', async (req: Request, res: Response) => {
  try {
    const user = await requireAuth(req);

    // Total jobs
    const totalJobs = await db.job.count({ where: { userId: user.id } });

    // Jobs today
    const today = new Date();
    today.setHours(0, 0, 0, 0);
    const jobsToday = await db.job.count({
      where: { userId: user.id, createdAt: { gte: today } }
    });

    // Done jobs count
    const doneJobs = await db.job.count({
      where: { userId: user.id, status: 'DONE' }
    });

    // Success rate
    const successRate = totalJobs > 0 ? Math.round((doneJobs / totalJobs) * 100) : 0;

    // Daily usage for last 30 days
    const thirtyDaysAgo = new Date();
    thirtyDaysAgo.setDate(thirtyDaysAgo.getDate() - 30);

    const jobs = await db.job.findMany({
      where: {
        userId: user.id,
        createdAt: { gte: thirtyDaysAgo }
      },
      select: { createdAt: true }
    });

    // Group by date
    const dailyMap = new Map<string, number>();
    for (const job of jobs) {
      const dateStr = job.createdAt.toISOString().split('T')[0];
      dailyMap.set(dateStr, (dailyMap.get(dateStr) || 0) + 1);
    }

    // Fill in missing dates with 0
    const dailyUsage = [];
    for (let i = 29; i >= 0; i--) {
      const date = new Date();
      date.setDate(date.getDate() - i);
      const dateStr = date.toISOString().split('T')[0];
      dailyUsage.push({ date: dateStr, count: dailyMap.get(dateStr) || 0 });
    }

    return res.json({
      credits: user.credits,
      totalJobs,
      jobsToday,
      successRate,
      dailyUsage,
    });
  } catch (err: any) {
    if (err.status) return res.status(err.status).json({ error: err.message });
    console.error(err);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// GET /billing/transactions
router.get('/transactions', async (req: Request, res: Response) => {
  try {
    const user = await requireAuth(req);

    const logs = await db.creditLog.findMany({
      where: { userId: user.id },
      orderBy: { createdAt: 'desc' },
      take: 50,
    });

    return res.json({ logs });
  } catch (err: any) {
    if (err.status) return res.status(err.status).json({ error: err.message });
    console.error(err);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// POST /billing/orders
router.post('/orders', async (req: Request, res: Response) => {
  try {
    if (!process.env.RAZORPAY_KEY_ID) {
      return res.status(503).json({ error: 'Billing not configured' });
    }

    const user = await requireAuth(req);

    const { pack } = req.body as { pack: PackType };
    if (!pack || !CREDIT_PACKS[pack]) {
      return res.status(400).json({ error: 'Invalid pack' });
    }

    const creditPack = CREDIT_PACKS[pack];
    const razorpay = new Razorpay({
      key_id: process.env.RAZORPAY_KEY_ID,
      key_secret: process.env.RAZORPAY_KEY_SECRET,
    });

    const order = await razorpay.orders.create({
      amount: creditPack.amount * 100,
      currency: 'INR',
      receipt: `order_${user.id}_${Date.now()}`,
      notes: { userId: user.id, pack },
    });

    return res.json({
      orderId: order.id,
      amount: order.amount,
      currency: order.currency,
      credits: creditPack.credits,
    });
  } catch (err: any) {
    if (err.status) return res.status(err.status).json({ error: err.message });
    console.error(err);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// POST /billing/verify
router.post('/verify', async (req: Request, res: Response) => {
  try {
    const user = await requireAuth(req);

    const { orderId, paymentId, signature } = req.body;
    if (!orderId || !paymentId || !signature) {
      return res.status(400).json({ error: 'Missing verification fields' });
    }

    const generatedSignature = crypto
      .createHmac('sha256', process.env.RAZORPAY_KEY_SECRET || '')
      .update(`${orderId}|${paymentId}`)
      .digest('hex');

    if (generatedSignature !== signature) {
      return res.status(400).json({ error: 'Invalid signature' });
    }

    const razorpay = new Razorpay({
      key_id: process.env.RAZORPAY_KEY_ID || '',
      key_secret: process.env.RAZORPAY_KEY_SECRET || '',
    });
    const order = await razorpay.orders.fetch(orderId);
    const pack = order.notes?.pack as PackType;
    const credits = CREDIT_PACKS[pack]?.credits || 500;

    const updated = await db.user.update({
      where: { id: user.id },
      data: { credits: { increment: credits } },
    });

    await db.creditLog.create({
      data: {
        userId: user.id,
        amount: credits,
        reason: 'purchase',
        jobId: orderId,
      },
    });

    return res.json({ ok: true, credits: updated.credits });
  } catch (err: any) {
    if (err.status) return res.status(err.status).json({ error: err.message });
    console.error(err);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;

File: api/routes/billing.py
from fastapi import APIRouter, Depends, Request
from sqlalchemy import select, func, case
from sqlalchemy.ext.asyncio import AsyncSession
from datetime import datetime, timedelta

from api.database import get_db
from api.models import User, Job, CreditLog
from api.deps import get_current_user
from api.schemas import (
    PlanSchema,
    BillingPurchaseResponse,
    BillingPurchaseRequest,
    BillingUsageResponse,
    BillingTransactionsResponse,
    CreditLogSchema,
)
from slowapi import Limiter
from slowapi.util import get_remote_address

router = APIRouter()
limiter = Limiter(key_func=get_remote_address)

PLANS = [
    PlanSchema(pack="starter", credits=500, amount=199, currency="INR", name="Starter Pack"),
    PlanSchema(pack="pro", credits=2000, amount=599, currency="INR", name="Pro Pack"),
    PlanSchema(pack="enterprise", credits=10000, amount=1999, currency="INR", name="Enterprise Pack"),
]


@router.get("/plans", response_model=list[PlanSchema])
async def get_plans():
    return PLANS


@router.post("/purchase", response_model=BillingPurchaseResponse)
@limiter.limit("5/minute")
async def purchase_credits(
    request: Request,
    body: BillingPurchaseRequest,
    current_user: User = Depends(get_current_user),
):
    return BillingPurchaseResponse(ok=True, message="Payment integration coming soon")


@router.get("/usage", response_model=BillingUsageResponse)
@limiter.limit("30/minute")
async def get_usage(
    request: Request,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    today_start = datetime.utcnow().replace(hour=0, minute=0, second=0, microsecond=0)

    agg_result = await db.execute(
        select(
            func.count().label("total"),
            func.sum(case((Job.status == "DONE", 1), else_=0)).label("done"),
            func.sum(case((Job.created_at >= today_start, 1), else_=0)).label("today"),
        ).where(Job.user_id == current_user.id)
    )
    row = agg_result.one()
    total_jobs = row.total or 0
    done_count = row.done or 0
    jobs_today = row.today or 0
    success_rate = (done_count / total_jobs * 100) if total_jobs > 0 else 0.0

    thirty_days_ago = datetime.utcnow() - timedelta(days=30)

    daily_query = await db.execute(
        select(
            func.date(Job.created_at).label("date"),
            func.count(Job.id).label("count"),
        )
        .where(
            Job.user_id == current_user.id,
            Job.created_at >= thirty_days_ago,
            Job.status == "DONE",
        )
        .group_by(func.date(Job.created_at))
        .order_by(func.date(Job.created_at))
    )
    daily_counts = {str(row.date): row.count for row in daily_query.all()}

    daily_usage = []
    for i in range(30):
        day = (datetime.utcnow() - timedelta(days=29 - i)).replace(hour=0, minute=0, second=0, microsecond=0)
        date_str = day.strftime("%Y-%m-%d")
        daily_usage.append({"date": date_str, "count": daily_counts.get(date_str, 0)})

    return BillingUsageResponse(
        credits=current_user.credits,
        totalJobs=total_jobs,
        jobsToday=jobs_today,
        successRate=round(success_rate, 2),
        dailyUsage=daily_usage,
    )


@router.get("/transactions", response_model=BillingTransactionsResponse)
async def get_transactions(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(CreditLog)
        .where(CreditLog.user_id == current_user.id)
        .order_by(CreditLog.created_at.desc())
        .limit(100)
    )
    logs = result.scalars().all()
    return BillingTransactionsResponse(logs=[CreditLogSchema.model_validate(l) for l in logs])